### DISCUSSION TITLE
is it possible to divide the model file in two files, one containing the a model and the other a reference to that model?

### DISCUSSION THREAD

**Discussion by gborghesan** [textX/textX#330] [Q&A]
Iam considering a case that is analogous to the entity-person example. 
I am looking at a language for robot programming with skills

I would like to have a metamodel of the Skills** and Skill Types in a file (**tasks_specification.tx**)

    Skill: 'skill' name=MY_ID '{'
      'type' '=' type=[SkillType]
      ('properties' '{' properties+=Property '}')?
    '}'
    ;
    SkillType: 'skill_type' name=MY_ID;
    Property:  name=MY_ID ':' type=BASETYPE ;

and then the definitions of various skill metamodel in *one or more files* (programmed by expert users) **assembly.tasks, screwing.tasks, etc...**

    skill_type PEG_IN_THE_HOLE
    skill_type GRASP_OBJ
    ...

and lastly the model of a skill (and the program itself) on a third file (**assembly_car1.prog**):

    skill PEG_IN_THE_HOLE_1 {
	   type = PEG_IN_THE_HOLE
	   properties{max_wrench : 10}
    }

1) It is possible to do that ?
2) is there a way to define a kind of "import" command in the last file (assembly_car1.prog) that points to the file that specifies the skill template ? 

Cheers, Gianni.

**addendum:** i see that the scoping could help in at least point 1, but i fail to understand how to use it, so far.

**Comment by igordejanovic** [comment]
Here is one way how to do it:

- task_specification.tx
```
Model:
    skills*=Skill
    skill_types*=SkillType
;

Skill: 'skill' name=ID '{'
    'type' '=' type=[SkillType]
    ('properties' '{' properties+=Property '}')?
'}'
;

SkillType: 'skill_type' name=ID;
Property:  name=ID ':' value=BASETYPE ;
```

- assembly.tasks
```
skill_type PEG_IN_THE_HOLE
skill_type GRASP_OBJ
```
- assembly_car1.prog
```
skill PEG_IN_THE_HOLE_1 {
   type = PEG_IN_THE_HOLE
   properties {max_wrench : 10}
}
```

And finally the test that show how it is loaded:
```python
from textx import metamodel_from_file
from textx.scoping import providers

mm = metamodel_from_file('task_specification.tx')

global_repo_provider = providers.PlainNameGlobalRepo()

# Register models to search during resolving
global_repo_provider.register_models('*.tasks')

mm.register_scope_providers({'Skill.type': global_repo_provider})

model = mm.model_from_file('assembly_car1.prog')

# Now lets verify that we have a properly loaded model
assert len(model.skills) == 1
skill = model.skills[0]
assert skill.name == 'PEG_IN_THE_HOLE_1'
assert skill.type.name == 'PEG_IN_THE_HOLE'
assert len(skill.properties) == 1
assert skill.properties[0].name == 'max_wrench'
assert skill.properties[0].value == 10

```

There are other ways. Maybe @goto40 could suggest some better way. :)

**Comment by goto40** [comment]
Hi!
The example given above is perfect!

Basically, you have four build-in choices for the problem (see http://www.igordejanovic.net/textX/scoping/), depending on the name lookup you wish (global names "MyName" of fully qualified names "namespace1.namespace2.MyName") and the way you make other model parts from other files visible in your model (implicit include "*Global*" or explicit include "*ImportURI*"):

    textx.scoping.providers.FQNGlobalRepo (decorated scope provider)
    textx.scoping.providers.PlainNameGlobalRepo (decorated scope provider)
    FQNImportURI (decorated scope provider)
    PlainNameImportURI (decorated scope provider)

The second one is the one used in the example given above.

You can see examples in the unittests referenced in http://www.igordejanovic.net/textX/scoping/ and http://www.igordejanovic.net/textX/multimetamodel/.

Do not hesitate to ask further questions...
Best wishes,
goto40

**Comment by igordejanovic** [comment]
I forgot to mention that these features are available in the development version only (`master` branch), i.e. not released yet.

**Comment by gborghesan** [comment]
dears, sorry for late answers (other deadlines incoming)
thanks for the very exhaustive answer
i did not delve into the addendum of @goto40 yet, but i have a comment/question on the example:
The one posted is of course a MWE, the metamodel has more syntax, so, once i added something like 
```
Model:
    '{' 'skills' skills*=Skill '}'
    skill_types*=SkillType
;
```
it did not work anymore. My understanding from the errors is that both files  
- `assembly.tasks`
- `assembly_car1.prog`
should comply to `task_specification.tx` (please confirm).
To do so i opted for the following for 
- `task_specification.tx`
```
Model:
    ('skills' '{'  skills+=Skill '}')|
    (skill_types+=SkillType)
;

Skill: 'skill' name=ID '{'
    'type' '=' type=[SkillType]
    ('properties' '{' properties+=Property '}')?
'}'
;

SkillType: 'skill_type' name=ID;
Property:  name=ID ':' value=BASETYPE ;
```
- `assembly_car1.prog`
```
skills{
skill PEG_IN_THE_HOLE_1 {
   type = PEG_IN_THE_HOLE
   properties {max_wrench : 10}
}
}
```
Practically, i have a submodel for each file. I am open to any suggestion on how to make it better.
Thanks again, Gianni.

**Comment by goto40** [comment]
Sorry. I was not online last week. I will check your comment the next
days...

Gianni <notifications@github.com> schrieb am Do., 24. Mai 2018, 12:22:

**Comment by goto40** [comment]
Hi.
I am not sure if I understood the correct aspect of your question - so do not hesitate to give me a hint what aspect you would like to have clarified in more detail.

For what I see you have multiple options:
 * You can use your solution where you have one single meta model for every model file. In this case all model files must conform to the meta model (***I confirm your comment***).
 * You could also use one meta model for the master model (e.g. for *.prog files) which includes another model (the *.task files). An example is found in the tests directory: tests/functional/test_scoping/metamodel_provider2
   * Here, a Base.tx grammar defines all relevant Types
   * Specific grammars using the Base.tx grammar then define different file types, e.g. containing ingredients (like your *.task) or recipes (like your *.prog). See: tests/functional/test_scoping/test_metamodel_provider2.py how to setup the meta models.
    * Of course every file must conform to a meta model grammar: however in this example *.task and *.prog will have a different syntax (which I think is a better solution).

**Comment by goto40** [comment]
See https://github.com/igordejanovic/textX/blob/master/docs/scoping.MD how to split the model across files (e.g., using the ImporURI class) . 

If you wish to reference elements from other models with different meta models, have a look at https://github.com/igordejanovic/textX/blob/master/docs/multimetamodel.md.

Is your problem solved with this?
